### PR TITLE
Demo for underconstrained inputs and outputs

### PR DESCRIPTION
**Pull request by 0xddom**

### CODE REVIEW

File: lib/analysis/taint-analysis/TaintAnalysis.h [hunk 1]
```diff
@@ -308,12 +308,25 @@ namespace vanguard {
         Observer *observer;
     };
 
+#define COLLECT_VALUES 0
```
Comment (kferles) [inline lib/analysis/taint-analysis/TaintAnalysis.h:311]: Why are we using macros? What is the expected value for this? Should we update the build instructions for things to work properly?

File: lib/detectors/UnderConstraints/UnderConstrainedArrayInput.h [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+            for (auto *unit : universe.units())
+                for (auto *component : unit->components())
+                    detectInComponent(component);
+        }
+
+        void report() override {
+            if (underConstrainedInputSignals.empty()) {
+                std::cout << "Nothing to report" << std::endl;
+            }
+            for (const auto& [component, outputSignals] : underConstrainedInputSignals) {
```
Comment (kferles) [inline lib/detectors/UnderConstraints/UnderConstrainedArrayInput.h:25]:
```suggestion
            for (const auto& [component, inputSignal] : underConstrainedInputSignals) {
```

File: lib/detectors/UnderConstraints/UnderConstrainedArrayInput.h [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+                std::cout << "In component " << component->name() << " the following input signals are not constrained:" << std::endl;
+                for (const auto *outputSignal : outputSignals) {
```
Comment (kferles) [inline lib/detectors/UnderConstraints/UnderConstrainedArrayInput.h:27]:
```suggestion
                for (const auto *inputSignal : inputSignals) {
```

File: lib/detectors/UnderConstraints/UnderConstrainedArrayInput.h [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+                    std::cout << "  - " << outputSignal->name() << std::endl;
```
Comment (kferles) [inline lib/detectors/UnderConstraints/UnderConstrainedArrayInput.h:28]:
```suggestion
                    std::cout << "  - " << inputSignal->name() << std::endl;
```

File: lib/detectors/UnderConstraints/UnderConstrainedArrayInput.h [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+            for (const auto& [key, _] : signalsInfo) {
+                inputSignals.push_back(key->unwrap());
+            }
+
+            TaintAnalysisResultsCollector gepsCollector;
+            TaintAnalysis gepsAnalysis(inputSignals, geps, &gepsCollector);
+            gepsAnalysis.run();
+            checkMissingIndexations(signalsInfo, gepsCollector, val2Signal, component);
+
+//            for (auto *constraint : component->getConstraints()) {
```
Comment (kferles) [inline lib/detectors/UnderConstraints/UnderConstrainedArrayInput.h:202]: Stray comments.

File: lib/detectors/UnderConstraints/UnderConstrainedOutput.h [hunk 6]
```diff
@@ -7,6 +7,14 @@
 
 #include "../UniverseDetector.h"
 
+template<> struct [[maybe_unused]] std::equal_to<SVF::VFGNode> {
+    bool operator()(const SVF::VFGNode *lhs, const SVF::VFGNode *rhs) const
+    {
+        std::cout << "Comparing " << lhs << " with " << rhs << std::endl;
```
Comment (kferles) [inline lib/detectors/UnderConstraints/UnderConstrainedOutput.h:13]:
```suggestion
        
```

File: lib/detectors/UnderConstraints/UnderConstrainedOutput.h [hunk 7]
```diff
@@ -93,28 +109,64 @@ namespace vanguard {
             outputsAnalysis.run();
 
             for (const auto& [outputSignal, connectedConstraints] : collector) {
-                if (connectedConstraints.empty()) {
+                if (connectedConstraints.sinks.empty()) {
                     // Output is disconnected from constraints.
-                    llvm::dbgs() << "Signal " << val2Signal.at(outputSignal)->name() << " is not connected to any constraint!\n";
+                    std::cout << "Signal " << val2Signal.at(outputSignal)->name() << " is not connected to any constraint!\n";
```
Comment (kferles) [inline lib/detectors/UnderConstraints/UnderConstrainedOutput.h:114]:
```suggestion
                    llvm::dbgs() << "Signal " << val2Signal.at(outputSignal)->name() << " is not connected to any constraint!\n";
```

File: lib/detectors/UnderConstraints/UnderConstrainedOutput.h [hunk 8]
```diff
@@ -93,28 +109,64 @@ namespace vanguard {
...
                     underConstrainedOutputSignals[component].push_back(val2Signal.at(outputSignal));
                 } else {
+                    std::cout << "Signal " << val2Signal.at(outputSignal)->name() << " is connected to some constraint! Is this constraint using inputs?\n";
```
Comment (kferles) [inline lib/detectors/UnderConstraints/UnderConstrainedOutput.h:117]: Is this for debugging or part of the output?

File: lib/detectors/UnderConstraints/UnderConstrainedOutput.h [hunk 9]
```diff
@@ -93,28 +109,64 @@ namespace vanguard {
...
+                    }
+                    if (underConstrOutput) {
+                        std::cout << "Signal " << val2Signal.at(outputSignal)->name() << " is not connected to any input signal!\n";
+                        underConstrainedOutputSignals[component].push_back(val2Signal.at(outputSignal));
+                        continue;
+                    }
+                    bool writesIntoOutputInSlice = false;
+                    for (const auto& [inputSignal, result] : inputsCollector) {
+                        if (result.sinks.empty()) continue;
+                        std::cout << "Signal " << val2Signal.at(outputSignal)->name() << " is connected to an input signal!\n";
```
Comment (kferles) [inline lib/detectors/UnderConstraints/UnderConstrainedOutput.h:144]: Is this part of the detector's output? If yes, is it necessary?

**Review by kferles — CHANGES_REQUESTED** [review]
Have some comments on some obvious things. We will clean up the logic in the next iteration with the new front-end.

**Review by 0xddom — COMMENTED** [review]
Rename variable names

File: nix/circom2llvm.nix [hunk 10]
```diff
@@ -8,13 +8,13 @@
 
 rustPlatform.buildRustPackage rec {
   pname = "circom2llvm";
-  version = "0.0.0";
+  version = "0.0.1";
```
[the remark this answers is not in the document]
Comment (0xddom) [reply nix/circom2llvm.nix:11]: might as well try

File: test/circom/uc-comp-out-bug.circom [hunk 11]
```diff
@@ -1,6 +1,6 @@
 pragma circom 2.0.0;
 // REQUIRES: circom2llvm
-// RUN: vanguard_driver --detector uc-comp-out -D %S %s | FileCheck --check-prefix UCOUT %s
+// RUN: vanguard_driver -D %S --detector uc-comp-out %s | FileCheck --check-prefix UCOUT %s
 // UCOUT: Component Bad does not constraint the signal LessThan.out
 
 // XFAIL: *
```
[the remark this answers is not in the document]
Comment (0xddom) [reply test/circom/uc-comp-out-bug.circom:6]: Lemme check. Also, does it really matter the order of the arguments in the driver?

[the remark this answers is not in the document]
Comment (Technius) [reply test/circom/uc-comp-out-bug.circom:6]: Nope, order doesn't matter.

[the remark this answers is not in the document]
Comment (0xddom) [reply test/circom/uc-comp-out-bug.circom:6]: yep, still crashes....

**Pull request merged**